### PR TITLE
Muelu tpetra dep

### PR DESCRIPTION
**Pull request by bathmatt**
@trilinos/muelu 
this is a lot of changes but have muelu working with the new tpetra interface and have many UVM issues fixed.  

This doesn't fix all of the UVM issues but it fixes a lot so I want to get this going and merged once the zoltan2 one gets fixes.

### CODE REVIEW

**Review by jhux2 — APPROVED** [review]
Approving to let the AT run.

**Comment by bathmatt** [comment]
@cgcgcg @jhux2  this is failing here

```
#13 0x00000000119960e4 in MueLu::NotayAggregationFactory<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >::localSpGEMM (this=0x40a6cbd0, A=..., B=..., 
    matrixLabel=..., C=...) at ../packages/muelu/src/Graph/PairwiseAggregation/MueLu_NotayAggregationFactory_def.hpp:913

```

It is calling into Kokkos Kernels here

```
    // Symbolic multiplication
    KokkosSparse::Experimental::spgemm_symbolic(&kh, A.numRows(),
                                                B.numRows(), B.numCols(),
                                                A.graph.row_map, A.graph.entries, false,
                                                B.graph.row_map, B.graph.entries, false,
                                                rowPtrC);

```

What kind of checks can i do outside of KK to get more info on why this is failing?  

I'm stuck a bit on how to proceed on this bug.

**Comment by cgcgcg** [comment]
@lucbv

**Comment by bathmatt** [comment]
new info, cuda-memcheck gets the test to pass, without cuda-memcheck it fails,

tested cuda-launch-blocking it fails in the same way

**Comment by lucbv** [comment]
It seems to fail because of an access out of bounds, it is strange that this would happen because of the refactor for UVM removal though.
I guess turning on Kokkos' bound checking makes sense otherwise all other options are painful and will revolve around printf...
I could take a look at it but not before tomorrow.

**Comment by bathmatt** [comment]
Really weird thing is that without UVM it passes as well...  I have a theory I'm testing but I don't believe this is something I caused, but has been floating around for a while.

**Comment by lucbv** [comment]
Yes, that would not be too surprising, this Factory is fairly experimental and is not used in production so it could have issues that have not been discovered yet...

**Comment by bathmatt** [comment]
Running bisect on my changes, will have a commit which triggered this error

**Comment by bathmatt** [comment]
It was this commit that caused the error, I'm going to look a bit more into this

d0cf4a57ddc99aa7ac7657754ca377537e833491

**Comment by bathmatt** [comment]
IT was this commit d0cf4a57ddc99aa7ac7657754ca377537e833491

can someone have a quick looksee to see if there is anything obvious,

**Comment by bathmatt** [comment]
I think I've found it, it is the parallel for not doing what I thought it should, Let me fix my code,

**Comment by lucbv** [comment]
You know the whole thing is pretty experimental still, we are not sure that we like this style of aggregation yet...
It could be left as is, without the parallel_for if it's a problem to fix it.

**Comment by bathmatt** [comment]
OK, this test passes with and without UVM

File: packages/zoltan2/core/src/TpetraCrsColorer/Zoltan2_TpetraCrsColorer.hpp
Comment (lucbv) [inline packages/zoltan2/core/src/TpetraCrsColorer/Zoltan2_TpetraCrsColorer.hpp:325]: Technically `local_matrix.graph` is cheaper as it will not recreate the local graph

**Review by lucbv — COMMENTED** [review]
Overall this looks good but I think a few things can be improved for future maintenance.

File: packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp
Comment (cgcgcg) [inline packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp:207]: Wouldn't write-only be enough here?

File: packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp
Comment (cgcgcg) [inline packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp:232]: Isn't read-only enough?

File: packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp
Comment (cgcgcg) [inline packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp:1790]: Write-only?

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_kokkos_def.hpp
Comment (bathmatt) [reply packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_kokkos_def.hpp:604]: So, the problem is that this code goes between requiring device and host copies of the matrix, you can't get host copies when the device one is active therefore I have to free kokkosMatrix and regrab the matrix.  I'd be interested in other ways to do this

Comment (bathmatt) [reply packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_kokkos_def.hpp:604]: @lucbv I reduced the scope of the kokkosMatrix reset and reinitialize, the ones below are gone with the latest patch

File: packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp
Comment (bathmatt) [reply packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp:1790]: I was not sure if all values are written or if some are just assumed to remain 0, can you confirm that 0s are written?

File: packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp
Comment (bathmatt) [reply packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp]: Fixed

File: packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp
Comment (bathmatt) [reply packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp:207]: Same as below, i am not sure that myColsToZero are 100% overwritten, can you confirm that it is.  I don't believe we can guarantee that initial values are 0

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_kokkos_def.hpp
Comment (lucbv) [reply packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_kokkos_def.hpp]: Thanks for having a look!

**Comment by bathmatt** [comment]
crap, fat fingered something, I'll push a fix, 

@jhux2 thoughts on @lucbv comment above?

For the zoltan fixes I'm going to push a zoltan2 only branch soon and then address these in this MR.

I don't want to rock that boat with the approval process.

File: packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp
Comment (cgcgcg) [reply packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp:1790]: Everything should be initialized to 0 when the MV is created here: https://github.com/trilinos/Trilinos/blob/d8c96d1ea19b5e60c1c6bc1b12c311f8b29b0ca0/packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp#L1787
Entries that aren't modified are supposed to stay zero.

**Comment by bathmatt** [comment]
@lucbv @cgcgcg I've addressed most the issues, and I've rebased off origin/develop.  there are some questions in the PR in the code segments to finish this up.  I think we can open another issue to address @lucbv comment and create a device version of GetDiagonalValues and use that so we wouldn't need to scope the kokkosMatrix life after this merges. 

Thoughts?  We can include it with this MR, but I want to make sure this doesn't go stale.

**Review by lucbv — APPROVED** [review]
@bathmatt thanks for looking into our questions.
I am fine with splitting things in multiple PRs although we might want to keep track of some of these future changes in an issue otherwise we might forget...

**Review by cgcgcg — APPROVED** [review]
Let's merge this. We can polish details in another PR.

**Pull request merged**
